File: app/src/main/java/org/anvei/novelreader/widget/read/BaseReadView.java
package org.anvei.novelreader.widget.read;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Scroller;

import java.util.ArrayList;
import java.util.List;

public abstract class BaseReadView<E extends View> extends ViewGroup {

    private static final String TAG = "BaseReadView";

    // 弹性滑动对象，实现过渡效果的滑动
    private final Scroller scroller;

    private final List<OnFlipOverListener> onFlipOverListenerList = new ArrayList<>();

    // 商定这个滑动是否有效的距离
    private static final int minScrollDistance = 0;
    // 翻页动画时间
    private static final int flipOverTime = 250;

    private final int screenWidth;


    private float startX = 0;
    private float curX = 0;

    private Direction initialDirection = Direction.NONE;    // 初始滑动方向
    private Direction curDirection = Direction.NONE;        // 实时滑动方向
    private Direction endDirection = Direction.NONE;        // 最终页面的滑动方向
    // 滑动的view
    private E scrolledView;

    protected E preView;
    protected E curView;
    protected E nextView;

    public BaseReadView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initChildView();
        scroller = new Scroller(context);
        screenWidth = context.getResources().getDisplayMetrics().widthPixels;
    }

    // 在这里完成子view的初始化
    protected void initChildView() {
        this.nextView = createChildView();
        this.curView = createChildView();
        this.preView = createChildView();
        addView(nextView);
        addView(curView);
        addView(preView);
        // 默认将最上层的view滑动到边缘（用于查看上一页）
        preView.scrollTo(-screenWidth, 0);
    }

    protected abstract E createChildView();

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int widthSize = MeasureSpec.getSize(widthMeasureSpec);
        int heightSize = MeasureSpec.getSize(heightMeasureSpec);
        // 设置view的测量大小
        setMeasuredDimension(widthSize, heightSize);
        for (int i = 0; i < getChildCount(); i++) {
            // 子view的MeasureSpec和FlipperLayout相同
            getChildAt(i).measure(widthMeasureSpec, heightMeasureSpec);
        }
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        for (int i = 0; i < getChildCount(); i++) {
            View child = getChildAt(i);
            int height = child.getMeasuredHeight();
            int width = child.getMeasuredWidth();
            // 子view全部叠放在一起，但是最顶层的子view被设置了scrollX，所以滑出了屏幕
            child.layout(0, 0, width, height);
        }
    }

    @Override
    public boolean performClick() {
        return super.performClick();
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        if (!isScrolling() && ev.getAction() == MotionEvent.ACTION_DOWN) {
            startX = ev.getX();
        }
        return super.dispatchTouchEvent(ev);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        performClick();
        if (isScrolling()) {
            return super.onTouchEvent(event);
        }
        switch (event.getAction()) {
            case MotionEvent.ACTION_MOVE:
                float oldX = curX;
                curX = event.getX();
                // 计算滑动距离，先确定本系列事件中最初的滑动方向，并确定被拖动的子view
                final float distance = startX - curX;
                if (initialDirection == Direction.NONE) {
                    if (distance > 0 && hasNextPage()) {
                        initialDirection = Direction.TO_LEFT;
                        scrolledView = curView;
                    } else if (distance < 0 && hasPrePage()) {
                        initialDirection = Direction.TO_RIGHT;
                        scrolledView = preView;
                    } else {
                        break;
                    }
                }
                // 确定本次MOVE事件相对于上次MOVE事件的滑动方向
                float relatedDistance = oldX - curX;
                if (relatedDistance > 0) {
                    curDirection = Direction.TO_LEFT;
                } else if (relatedDistance < 0) {
                    curDirection = Direction.TO_RIGHT;
                }
                if (initialDirection == Direction.TO_LEFT) {
                    if (distance > 0) {
                        scrolledView.scrollTo((int) distance, 0);
                    }
                } else {
                    if (distance < 0) {
                        scrolledView.scrollTo(screenWidth + (int) distance, 0);
                    }
                }
                break;
            case MotionEvent.ACTION_UP:
                if (initialDirection != Direction.NONE && curDirection != Direction.NONE) {
                    int scrollX = scrolledView.getScrollX();
                    int dx;
                    // 最初的翻页方向和最终翻页方向不同，所以不应翻页，应当将视图复位
                    if (initialDirection != curDirection) {
                        endDirection = Direction.NONE;
                        if (initialDirection == Direction.TO_LEFT) {
                            dx = -scrollX;
                        } else {
                            dx = screenWidth - scrollX;
                        }
                    } else { // 完成翻页，并且加载新的视图
                        endDirection = initialDirection;
                        if (initialDirection == Direction.TO_LEFT) {
                            dx = screenWidth - scrollX;
                        } else {
                            dx = -scrollX;
                        }
                    }
                    scroller.startScroll(scrollX, 0, dx, 0, flipOverTime);
                    postInvalidate();
                    initialDirection = Direction.NONE;
                    curDirection = Direction.NONE;
                }
                break;
        }
        return true;
    }

    @Override
    public void computeScroll() {
        super.computeScroll();
        if (scroller.computeScrollOffset()) {
            scrolledView.scrollTo(scroller.getCurrX(), scroller.getCurrY());
            postInvalidate();
        } else if (endDirection != Direction.NONE) {
            if (endDirection == Direction.TO_LEFT) {
                for (OnFlipOverListener onFlipOverListener : onFlipOverListenerList) {
                    onFlipOverListener.onNext();
                }
                removeView(preView);
                E cacheView = preView;
                preView = curView;
                curView = nextView;
                nextView = getView(cacheView, endDirection);
                if (nextView != null) {
                    nextView.scrollTo(0, 0);
                    addView(nextView, 0);
                }
            } else {
                for (OnFlipOverListener onFlipOverListener : onFlipOverListenerList) {
                    onFlipOverListener.onPre();
                }
                removeView(nextView);
                E cacheView = nextView;
                nextView = curView;
                curView = preView;
                preView = getView(cacheView, endDirection);
                if (preView != null) {
                    preView.scrollTo(-screenWidth, 0);
                    addView(preView);
                }
            }
            endDirection = Direction.NONE;
        }
    }

    protected abstract boolean hasNextPage();

    protected abstract boolean hasPrePage();

    protected abstract E getView(E cacheView, Direction direction);

    public void addOnFlipOverListener(OnFlipOverListener onFlipOverListener) {
        onFlipOverListenerList.add(onFlipOverListener);
    }

    public void removeFlipOverListener(OnFlipOverListener onFlipOverListener) {
        onFlipOverListenerList.remove(onFlipOverListener);
    }

    public void removeFlipOverListener(int index) {
        onFlipOverListenerList.remove(index);
    }

    // 翻页监听
    public interface OnFlipOverListener {

        void onNext();

        void onPre();
    }

    protected boolean isScrolling() {
        return !scroller.isFinished();
    }
}

File: readview/src/main/java/org/klee/readview/widget/BaseReadView.kt
/**
 * MIT License

 * Copyright (c) 2023 Klee

 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.klee.readview.widget

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.PointF
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.view.ViewConfiguration
import android.view.ViewGroup
import org.klee.readview.delegate.*
import org.klee.readview.utils.angle
import org.klee.readview.utils.apartFrom

/**
 * 负责提供多种翻页模式下的动画实现
 */
private const val TAG = "BaseReadView"
open class BaseReadView(context: Context, attributeSet: AttributeSet?)
    : ViewGroup(context, attributeSet) {

    internal val contentConfig by lazy { org.klee.readview.config.ContentConfig() }
    private val touchSlop = ViewConfiguration.get(context).scaledTouchSlop

    internal lateinit var curPageView: PageView
    internal lateinit var prePageView: PageView
    internal lateinit var nextPageView: PageView

    var shadowWidth: Int = 15

    private var onClickRegionListener: OnClickRegionListener? = null

    private var isMove = false

    val startPoint by lazy { PointF() }         // 第一次DOWN事件的坐标
    val lastPoint by lazy { PointF() }          // 上一次触摸的坐标
    val touchPoint by lazy { PointF() }         // 当前触摸的坐标
    private var initDirection =  PageDirection.NONE     // 本轮事件中最初的手势移动方向

    var flipMode = FlipMode.NoAnim
        set(value) {
            // 翻页模式改变了
            if (field != value && tempValue != null) {
                tempValue = createPageDelegate(value)
            }
            field = value
        }
    private var tempValue: PageDelegate? = null
    private val pageDelegate: PageDelegate get() {
        if (tempValue == null) {
            tempValue = createPageDelegate(flipMode)
        }
        return tempValue!!
    }

    /**
     * 创建一个PageDelegate
     */
    private fun createPageDelegate(mode: FlipMode): PageDelegate {
        return when (mode) {
            FlipMode.NoAnim ->
                NoAnimPageDelegate(this)
            FlipMode.Cover ->
                CoverPageDelegate(this)
        }
    }

    fun setPageDelegate(pageDelegate: PageDelegate) {
        this.tempValue = pageDelegate
    }

    /**
     * 对外提供的ReadPage自定义API函数，可以通过该函数配置PageView的内容视图、页眉视图、页脚视图
     * @param initializer 初始化器
     */
    open fun initPage(initializer: (pageView: PageView, position: Int) -> Unit) {
        curPageView = PageView(context)
        prePageView = PageView(context)
        nextPageView = PageView(context)
        initializer(curPageView, 0)
        initializer(prePageView, -1)
        initializer(nextPageView, 1)
        if (!(curPageView.initFinished && prePageView.initFinished && nextPageView.initFinished)) {
            throw IllegalStateException("没有完成PageView的初始化！")
        }
        curPageView.content.config = contentConfig
        prePageView.content.config = contentConfig
        nextPageView.content.config = contentConfig
        addView(nextPageView)
        addView(curPageView)
        addView(prePageView)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val widthSize = MeasureSpec.getSize(widthMeasureSpec)
        val heightSize = MeasureSpec.getSize(heightMeasureSpec)
        setMeasuredDimension(widthSize, heightSize)
        // 设置子view的测量大小
        for (i in 0 until childCount) {
            val child = getChildAt(i)
            if (child.visibility != GONE) {
                child.measure(widthMeasureSpec, heightMeasureSpec)
            }
        }
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        for (i in 0 until childCount) {
            val child = getChildAt(i)
            val height = child.measuredHeight
            val width = child.measuredWidth
            // 子view全部叠放在一起，但是最顶层的子view被设置了scrollX，所以滑出了屏幕
            child.layout(0, 0, width, height)
        }
        pageDelegate.onScrollValue()
    }

    private fun upTouchPoint(event: MotionEvent) {
        lastPoint.set(touchPoint)
        touchPoint.set(event.x, event.y)
    }

    private fun upStartPointer(x: Float, y: Float) {
        startPoint.set(x, y)
    }

    /**
     * 判断页面是否发生了翻页
     */
    private fun isPageMove(): Boolean {
        return (initDirection == PageDirection.PREV && hasPrevPage())
                || initDirection == PageDirection.NEXT && hasNextPage()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        upTouchPoint(event)
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                upStartPointer(event.x, event.y)
                pageDelegate.onTouch(event)
                isMove = false
                initDirection = PageDirection.NONE
            }
            MotionEvent.ACTION_MOVE -> {
                if (!isMove) {
                    // touchSlop确定是的isMove，该标记位用来区分点击和滑动事件
                    isMove = startPoint.apartFrom(touchPoint) > touchSlop
                    if (isMove) {
                        // 计算滑动的角度
                        val angle = startPoint.angle(touchPoint)
                        Log.d(TAG, "onTouchEvent: angle = $angle")
                        initDirection = pageDelegate.initDirection(angle)
                    }
                }
                if (isPageMove()) {
                    pageDelegate.onTouch(event)
                }
            }
            MotionEvent.ACTION_UP -> {
                Log.d(TAG, "onTouchEvent: startPoint = $startPoint, touchPoint = $touchPoint")
                if (!isMove) {
                    // 触发点击事件
                    val xPercent = startPoint.x / width * 100
                    val yPercent = startPoint.y / height * 100
                    if (!onClickRegion(xPercent.toInt(), yPercent.toInt())) {
                        // 如果onClickRegion()没有拦截该点击事件，就触发View的OnClickListener的回调
                        performClick()
                    }
                }
                if (isPageMove()) {
                    // 本系列事件为滑动事件，触发滑动动画
                    pageDelegate.onTouch(event)
                }
            }
        }
        return true
    }

    /**
     * 点击事件回调
     * @param xPercent 点击的位置在x轴方向上的百分比，例如xPercent=50，表示点击的位置为屏幕x轴方向上的中间
     * @param yPercent 点击的位置在y轴方向上的百分比
     * @return 表示是否拦截本次点击事件，如果onClickRegion()没有拦截该点击事件，就将本次点击事件转交OnClickListener处理
     */
    protected open fun onClickRegion(xPercent: Int, yPercent: Int): Boolean {
        return onClickRegionListener?.onClickRegion(xPercent, yPercent) ?: false
    }

    override fun computeScroll() {
        super.computeScroll()
        pageDelegate.computeScrollOffset()
    }

    override fun dispatchDraw(canvas: Canvas?) {
        super.dispatchDraw(canvas)
        canvas?.let {
            pageDelegate.dispatchDraw(canvas)
        }
    }

    open fun hasNextPage() = false

    open fun hasPrevPage(): Boolean = false

    protected open fun onFlipToPrev() = Unit

    protected open fun onFlipToNext() = Unit

    /**
     * 翻页完成以后，将会调用该函数进行子视图更新
     */
    internal open fun updateChildView(convertView: PageView,
                                      direction: PageDirection
    ): PageView {
        if (direction == PageDirection.PREV) {
            onFlipToPrev()
        }
        if (direction == PageDirection.NEXT) {
            onFlipToNext()
        }
        return convertView
    }

    /**
     * 通过函数调用来翻向下一页
     * @return 是否成功翻页
     */
    fun nextPage(): Boolean {
        if (hasNextPage()) {
            pageDelegate.nextPage()
            return true
        }
        return false
    }

    /**
     * 通过函数调用来翻向上一页
     * @return 是否成功翻页
     */
    fun prevPage(): Boolean {
        if (hasPrevPage()) {
            pageDelegate.prevPage()
            return true
        }
        return false
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        contentConfig.destroy()
    }

    interface OnClickRegionListener {
        /**
         * 点击事件回调
         * @param xPercent 点击的位置在x轴方向上的百分比，例如xPercent=50，表示点击的位置为屏幕x轴方向上的中间
         * @param yPercent 点击的位置在y轴方向上的百分比
         * @return 表示是否拦截本次点击事件
         */
        fun onClickRegion(xPercent: Int, yPercent: Int): Boolean
    }

    fun setOnClickRegionListener(listener: OnClickRegionListener) {
        this.onClickRegionListener = listener
    }

    fun setOnClickRegionListener(listener: (xPercent: Int, yPercent: Int) -> Boolean) {
        this.onClickRegionListener = object : OnClickRegionListener {
            override fun onClickRegion(xPercent: Int, yPercent: Int): Boolean {
                return listener(xPercent, yPercent)
            }
        }
    }

}